src/ignore.c

```diff
@@ -458,10 +458,25 @@ int git_ignore__lookup(
 	if (ignore_lookup_in_rules(out, ignores->ign_internal, &path))
 		goto cleanup;
 
-	/* next process files in the path */
-	git_vector_foreach(&ignores->ign_path, i, file) {
-		if (ignore_lookup_in_rules(out, file, &path))
-			goto cleanup;
+	/* next process files in the path.
+	 * this process has to consider the order that .ignore
+	 * files were discovered to ensure correct prioritization of rules
+	 */
+	if (ignores->flags & GIT_IGNORE_TOP_DOWN_DISCOVERY)
+	{
+		git_vector_rforeach(&ignores->ign_path, i, file) {
+			if (ignore_lookup_in_rules(out, file, &path)) {
+				goto cleanup;
+			}
+		}
```

Comment:
I think this makes sense to me. There cannot be an ignore rule that is invalidated by its parent rules, but the reverse can be true as demonstrated by your tests. The downside is that we need to always go all the way to the top for files that were ignored in the top-level directory, where we could previously immediately abort in the first iteration.

Naturally, correctness outweighs performance though.

Reply:
We could try to work out the starting `ignore` by taking a minimum of current depth or length of `ignores`. What do you think?